src/middlewares/createTelemetryMiddleware.ts

```diff
@@ -0,0 +1,81 @@
+import { Middleware } from '../types';
+
+type TelemetryWidget = {
+  type: string;
+  params: string[];
+  officialWidget: boolean;
+};
+
+export function createTelemetryMiddleware(): Middleware {
+  const isTelemetryEnabled =
+    typeof window !== 'undefined' &&
+    window.navigator.userAgent.indexOf('Algolia Crawler') > -1;
+
+  if (!isTelemetryEnabled) {
+    return () => ({
+      onStateChange() {},
+      subscribe() {},
+      unsubscribe() {},
+    });
+  }
+
+  const payload: { widgets: TelemetryWidget[] } = {
+    widgets: [],
+  };
+
+  const payloadContainer = document.createElement('meta');
+  const refNode = document.querySelector('head')!;
+  payloadContainer.name = 'instantsearch:widgets';
+  refNode.appendChild(payloadContainer);
+
+  return ({ instantSearchInstance }) => {
+    return {
+      onStateChange() {},
+      subscribe() {
+        setTimeout(() => {
+          const widgets = instantSearchInstance.mainIndex.getWidgets();
+
+          const parent = instantSearchInstance.mainIndex;
+
+          const initOptions = {
+            instantSearchInstance,
+            parent,
+            scopedResults: parent.getScopedResults(),
+            state: parent.getHelper()!.state,
+            helper: parent.getHelper()!,
+            createURL: parent.createURL,
+            uiState: instantSearchInstance._initialUiState,
+            renderState: instantSearchInstance.renderState,
+            templatesConfig: instantSearchInstance.templatesConfig,
+            searchMetadata: {
+              isSearchStalled: instantSearchInstance._isSearchStalled,
+            },
+          };
+
+          widgets.forEach(widget => {
+            const widgetParams = widget.getWidgetRenderState
+              ? (widget.getWidgetRenderState(initOptions) as any).widgetParams
```

Comment:
I think technically `widgetParams` returned from `getWidgetRenderState` can be nullish. For example, user implemented the function but returns null in some cases.

Reply:
ah that's true, and would cause a nasty error